### PR TITLE
Fix type of data_ptr in convert_aload/astore and useless typecasts

### PR DESCRIPTION
**Pull request by romainbrenguier**
This remove unnecessary typecasts in the case of baload and bastore of boolean arrays and fixes the type of the of the data pointer, and adds some unit test for convert_aload.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -2851,16 +2851,31 @@ code_blockt java_bytecode_convert_methodt::convert_ret(
   return c;
 }
 
+/// Add typecast if necessary to \p e to make it compatible with array type
+/// corresponding to \p type_char (see \ref java_array_type(const char)).
+/// Character 'b' is used both for `byte` and `boolean`.
+static exprt ensure_array_type(exprt e, char type_char)
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: :pick: `e` -> `expr`

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 2]
```diff
@@ -2851,16 +2851,31 @@ code_blockt java_bytecode_convert_methodt::convert_ret(
   return c;
 }
 
+/// Add typecast if necessary to \p e to make it compatible with array type
+/// corresponding to \p type_char (see \ref java_array_type(const char)).
+/// Character 'b' is used both for `byte` and `boolean`.
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: :question: Could we elaborate this last sentence as I'm struggling to match it up against the code. E.g. `'b' is used for both byte and boolean arrays, so if the array is a boolean and we're using a `b` operation then skip the cast`

:bulb: Perhaps the code could be simplified to be
```
const auto access_type = 
  ref_type && type_char == 'b' &&
    ref_type->subtype().get_identifier() == "java::array[boolean]" ? ref_type : java_array_type(type_char)
return typecast_exprt::conditional_cast(e, access_type);
```

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 1 above)
Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: :bulb: Function name - `conditional_array_cast`

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 3]
```diff
@@ -286,3 +288,84 @@ SCENARIO(
...
+{
+  const typet byte_array_type = java_array_type('b');
+  const symbol_exprt byte_array{"byte_array", byte_array_type};
+  const exprt offset = from_integer(1, java_int_type());
+  const exprt result = java_bytecode_convert_method_unit_testt::convert_aload(
+    "baload", {byte_array, offset});
+
+  // Result is of the form `(int)(*(byte_array->data + offset))`
+  auto query = make_query(result)
+                 .as<typecast_exprt>()[0]
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: :question: Why do we cast to int?

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 4]
```diff
@@ -286,3 +288,84 @@ SCENARIO(
...
+                 .as<dereference_exprt>()[0]
+                 .as<plus_exprt>();
+  REQUIRE(query[1].get() == offset);
+  REQUIRE(
+    query[0].as<member_exprt>()[0].as<dereference_exprt>()[0].get() ==
+    boolean_array);
+  // boolean_array->data has type *boolean
+  REQUIRE(
+    query[0].as<member_exprt>().get().type() ==
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: :question: This seems a little counter-intuitive since JVM stores boolean arrays as byte arrays, is there a good reason why we don't?

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 5]
```diff
@@ -286,3 +288,84 @@ SCENARIO(
...
+{
+public:
+  static exprt
+  convert_aload(const irep_idt &statement, const exprt::operandst &op)
+  {
+    return java_bytecode_convert_methodt::convert_aload(statement, op);
+  }
+};
+
+TEST_CASE(
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: :green_heart:

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 6]
```diff
@@ -369,3 +379,112 @@ TEST_CASE(
...
+            pointer_type(java_boolean_type()));
+        }
+      }
+    }
+    WHEN("iastore is called on the boolean array")
+    {
+      const code_blockt result =
+        java_bytecode_convert_method_unit_testt::convert_astore(
+          converter, "iastore", {boolean_array, offset, value}, location);
+      THEN("The result contains a typecast ")
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: :pick: The string as you put in the other `THEN` would be great!

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 7]
```diff
@@ -369,3 +379,112 @@ TEST_CASE(
...
+            plus[0].as<member_exprt>().get().type() ==
+            pointer_type(java_boolean_type()));
+        }
+      }
+    }
+    WHEN("iastore is called on the boolean array")
+    {
+      const code_blockt result =
+        java_bytecode_convert_method_unit_testt::convert_astore(
+          converter, "iastore", {boolean_array, offset, value}, location);
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:501]: :question: Is this even valid? Still good to test though :slightly_smiling_face:

**Review by thk123 — COMMENTED** [review]
A dream PR :+1: one blocking request for docs and a couple of questions. Also a TG bump would be good if you've not already done it

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 1 above)
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: done

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 2 above)
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: I've added the documentation, and simplified a bit the code, though not exactly like you suggested.

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp (hunk 3 above)
Comment (romainbrenguier) [reply jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: This is documented in `java_bytecode_promotion`, I've added a reference to that here.

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp (hunk 4 above)
Comment (romainbrenguier) [reply jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: The type of `boolean` is 8 bits so I think that enough to be compatible with the JVM. The reason we do this is to avoid some unecessary typecasts which then result into byte_extract operations when we attempt to retrieve a boolean from a boolean array that has been cast to a byte array.

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp (hunk 6 above)
Comment (romainbrenguier) [reply jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: done

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp (hunk 7 above)
Comment (romainbrenguier) [reply jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:501]: I guess it is not, according to the JVM documentation https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-6.html#jvms-6.5.iastore
the array must have int elements. I will add an invariant and see what happens.

Comment (romainbrenguier) [reply jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:501]: Adding an invariant does not work at the moment, because the array data is sometimes cast to void*. I'm working on another fix for that.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 053fef1defe331517e8c01dd04d87ea64c26e63c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117928447

**Review by thk123 — APPROVED** [review]
:green_heart:

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 8]
```diff
@@ -2851,16 +2851,33 @@ code_blockt java_bytecode_convert_methodt::convert_ret(
   return c;
 }
 
+/// Add typecast if necessary to \p expr to make it compatible with array type
+/// corresponding to \p type_char (see \ref java_array_type(const char)).
+/// Character 'b' is used both for `byte` and `boolean` arrays, so if \p expr
+/// is a boolean array an and we are using a `b` operation we can skip the
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: :pick: Redundant `an`

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 9]
```diff
@@ -2851,16 +2851,33 @@ code_blockt java_bytecode_convert_methodt::convert_ret(
...
+/// typecast.
+static exprt conditional_array_cast(exprt expr, char type_char)
```
Comment (allredj) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: Why not pass `expr` as a const ref?

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 10]
```diff
@@ -286,3 +288,86 @@ SCENARIO(
...
+  {
+    return java_bytecode_convert_methodt::convert_aload(statement, op);
+  }
+};
+
+TEST_CASE(
+  "baload byte array",
+  "[core][java_bytecode][java_bytecode_convert_method][convert_aload]")
+{
+  const typet byte_array_type = java_array_type('b');
```
Comment (allredj) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: :pick: I find it a bit hard to parse these tests. Maybe use `GIVEN`, `WHEN`, `THEN` as in the other ones?

**Review by allredj — APPROVED** [review]
Looks good.

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 9 above)
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: done

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp (hunk 8 above)
Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_bytecode_convert_method.cpp]: done

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp (hunk 10 above)
Comment (romainbrenguier) [reply jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp]: done

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: ac6b31e92ca568db9496b149d1a71419124ca0f9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118051176
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2e4099e27cba1aadcf1838636f4d78943b45c843).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118052198
